### PR TITLE
Add TypeScript definition

### PR DESCRIPTION
**Pull request by cancerberoSgx**
add typings and update dependencies

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for creating a TypeScript definition. 🙌 Can you follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide)?

**Comment by cancerberoSgx** [comment]
> Thanks for creating a TypeScript definition.  Can you follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide)?

OK, I will check and verify it. BTW - I mostly copy&paste it from p-map's.

**Comment by sindresorhus** [comment]
@BendingBender Can you help review?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * Returns a `Promise` that is fulfilled when all promises in `input` and ones returned from `filterer` are fulfilled, or rejects if any of the promises reject. The fulfilled value is an Array of the fulfilled values returned from filterer in input order.
+ *
+ * @param input - Iterated over concurrently in the `filterer` function.
+ * @param filterer - Function which is called for every item in `input`. Expected to return a `Promise` or value.
+ * @param options - Options-object.
+ *
+ * @example
+ *
+ * const pFilter = require('p-filter');
+ * const getWeather = require('get-weather'); // not a real module
```
Comment (sindresorhus) [inline index.d.ts:28]: Should use `import` statement as this is TS code here.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	concurrency?: number;
+}
+
+/**
+ * Function which is called for every item in `input`. Expected to return a `Promise` or value.
+ *
+ * @param input - Iterated element.
+ * @param index - Index of the element in the source array.
+ */
+export type Filterer<Element = any> = (input: Element, index: number) => boolean | Promise<boolean>
```
Comment (sindresorhus) [inline index.d.ts:16]:
```suggestion
export type Filterer<Element = any> = (input: Element, index: number) => boolean | Promise<boolean>;
```

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Returns a `Promise` that is fulfilled when all promises in `input` and ones returned from `filterer` are fulfilled, or rejects if any of the promises reject. The fulfilled value is an Array of the fulfilled values returned from filterer in input order.
+ *
+ * @param input - Iterated over concurrently in the `filterer` function.
+ * @param filterer - Function which is called for every item in `input`. Expected to return a `Promise` or value.
+ * @param options - Options-object.
```
Comment (sindresorhus) [inline index.d.ts:23]: No point in this since it just repeats the parameter name.

File: index.test-d.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import {expectType} from 'tsd-check';
+import pFilter from '.'
+
+expectType<Promise<string[]>>(pFilter(['foo'], s => !!s));
+expectType<string[]>(await pFilter(['foo', Promise.resolve('foo')], s => !!s, {concurrency: 1}));
+expectType<Promise<string[]>>(pFilter(['foo'], async s => !!s));
+expectType<Promise<number[]>>(pFilter([Promise.resolve(1)], n => !!n));
+expectType<Promise<string[]>>(pFilter(['foo'], () => false));
+expectType<Promise<number[]>>(pFilter([1, Promise.resolve(1)], async (n, i) => i === 0 ? !!n : Promise.resolve(!n)));
+expectType<Promise<number[]>>(pFilter([1, Promise.resolve(1)], async (n, i) => i === 0 ? !!n : Promise.resolve(!n), {concurrency: 2}));
```
Comment (sindresorhus) [inline index.test-d.ts:10]: Don't use one-character variable names or abbreviations.

File: index.test-d.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import {expectType} from 'tsd-check';
+import pFilter from '.'
+
+expectType<Promise<string[]>>(pFilter(['foo'], s => !!s));
```
Comment (sindresorhus) [inline index.test-d.ts:4]:
```suggestion
expectType<Promise<string[]>>(pFilter(['foo'], s => Boolean(s)));
```

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * };
+ * 
+ * (async () => {
+ * 	const result = await pFilter(places, filterer);
+ * 
+ * 	console.log(result);
+ * 	//=> ['Bangkok, Thailand']
+ * })();
+ */
+export default function <Element>(input: Iterable<Element | Promise<Element>>, filterer: Filterer<Element>, options?: Options): Promise<Element[]>;
```
Comment (sindresorhus) [inline index.d.ts:49]: The function should have a name.

File: index.d.ts (hunk 2 above)
Comment (BendingBender) [inline index.d.ts:16]: Please use `unknown` instead of `any`.

File: index.d.ts (hunk 6 above)
Comment (BendingBender) [inline index.d.ts:49]: `Element` will default to `any` when it can't be automatically determined, please use `<Element extends unknown>`

**Pull request closed without merging**